src/meta_ops.rs

```diff
@@ -735,7 +739,105 @@ pub fn less_equal<'gc>(
     lhs: Value<'gc>,
     rhs: Value<'gc>,
 ) -> Result<MetaResult<'gc, 2>, MetaOperatorError> {
-    meta_metaop(ctx, lhs, rhs, MetaMethod::Le, |a, b| {
+    meta_metaop(ctx, lhs, rhs, MetaMethod::Le, |_, a, b| {
         Some(a.to_constant()?.less_equal(&b.to_constant()?)?.into())
     })
 }
+
+/// The result of a concat metaoperation, either a completed [`Value`]
+/// or a [`Function`] that must be called with the values to
+/// concatenate.
+#[derive(Debug, Clone, Collect)]
+#[collect(no_drop)]
+pub enum ConcatMetaResult<'gc> {
+    Value(Value<'gc>),
+    Call(Function<'gc>),
+}
+
+/// Returns an estimate of the length of the concatenation of a list of values,
+/// or returns [`None`] if any value is not implicitly coercible to a string.
+pub fn estimate_concatenated_len<'gc>(
+    values: &[Value<'gc>],
+) -> Result<Option<usize>, MetaOperatorError> {
+    let mut len = 0usize;
+    for value in values {
+        let value_len = match value {
+            Value::Integer(i) => i.ilog10() as usize + i.is_negative() as usize,
+            Value::Number(_n) => 10,
+            Value::String(s) => s.as_bytes().len(),
+            _ => return Ok(None),
+        };
+        len = len
+            .checked_add(value_len)
+            .ok_or(MetaOperatorError::ConcatOverflow)?;
+    }
+    Ok(Some(len))
+}
```

Comment:
This is so hyper specific that I would choose any other option than have this as a public method.

I would rather just repeat the code, or even just not estimate length at all.

Reply:
Ah, whoops -- I didn't mean to make that fully public.  Would it be reasonable to make it `pub(crate)` instead, or should I just put it back into the two concat impls?  It's the simplest part to share between the two versions.

For now I'll make it `pub(crate)`, but I can inline it again if needed.  Though it may not be obvious that changes to this would also need to be reflected in `table.rs`... would it be better to move all of the concatenation code into `meta_ops.rs`, even the stuff that handles separators?

Reply:
`estimate_concatenated_len` is no longer public, and I ended up moving the separated version into `meta_ops`, next to the normal version.  (It may not be the best place for it, since there's no meta-operator that uses it, but it's better to have the two variants of concat in the same place.)